contracts/test/psm/PegStabilityModuleNonCustodial.t.sol

```diff
@@ -0,0 +1,235 @@
+      uint256 endingPSMUnderlyingBalance = psm.balance();
+      uint256 endingPCVDepositUnderlyingBalance = underlyingToken.balanceOf(address(pcvDeposit));
+
+      assertEq(endingPCVDepositUnderlyingBalance, mintAmount * 2);
+      assertEq(endingPSMUnderlyingBalance, 0);
+      assertEq(endingUserFEIBalance, mintAmount * 2);
+    }
+
+    /// @notice pcv deposit gets depleted on redeem
+    function testSwapFeiForUnderlying() public prank(userAddress) {
+      fei.approve(address(psm), mintAmount);
+      psm.redeem(userAddress, mintAmount, mintAmount);
+
+      uint256 endingUserFEIBalance = fei.balanceOf(userAddress);
+      uint256 endingUserUnderlyingBalance = underlyingToken.balanceOf(userAddress);
+      uint256 endingPSMUnderlyingBalance = psm.balance();
+      uint256 endingPCVDepositUnderlyingBalance = underlyingToken.balanceOf(address(pcvDeposit));
+
+      assertEq(endingPSMUnderlyingBalance, 0);
+      assertEq(endingUserFEIBalance, 0);
+      assertEq(endingUserUnderlyingBalance, mintAmount * 2);
+      assertEq(endingPCVDepositUnderlyingBalance, 0);
+    }
+
+    /// @notice redeem fails when price is out of bounds
+    function testMintFailsAboveCeiling() public prank(userAddress) {
+      oracle.setExchangeRate(2);
+      vm.expectRevert(
+          abi.encodePacked("PegStabilityModule: price out of bounds")
+      );
+
+      psm.mint(userAddress, mintAmount, mintAmount);
+    }
+
+    /// @notice redeem fails when price is out of bounds
+    function testMintFailsBelowFloor() public prank(userAddress) {
+      oracle.setExchangeRateScaledBase(95e16);
+      vm.expectRevert(
+          abi.encodePacked("PegStabilityModule: price out of bounds")
+      );
+
+      psm.mint(userAddress, mintAmount, mintAmount);
+    }
+
+    /// @notice redeem fails when price is out of bounds
+    function testRedeemFailsBelowFloor() public prank(userAddress) {
+      oracle.setExchangeRateScaledBase(95e16);
+      vm.expectRevert(
+          abi.encodePacked("PegStabilityModule: price out of bounds")
+      );
+
+      psm.redeem(userAddress, mintAmount, mintAmount);
+    }
+
+    /// @notice redeem fails when price is out of bounds
+    function testRedeemFailsAboveCeiling() public prank(userAddress) {
+      oracle.setExchangeRate(2);
+      vm.expectRevert(
+          abi.encodePacked("PegStabilityModule: price out of bounds")
+      );
+
+      psm.redeem(userAddress, mintAmount, mintAmount);
+    }
+
+    /// @notice redeem fails without approval
+    function testSwapFeiForUnderlyingFailsWithoutApproval() public prank(userAddress) {
+      vm.expectRevert(
+          abi.encodePacked("ERC20: transfer amount exceeds allowance")
+      );
+
+      psm.redeem(userAddress, mintAmount, mintAmount);
+    }
+
+    /// @notice mint fails without approval
+    function testSwapUnderlyingForFeiFailsWithoutApproval() public prank(userAddress) {
+      vm.expectRevert(
+          abi.encodePacked("ERC20: transfer amount exceeds allowance")
```

Comment:
Surely you're expecting a different error message to this if a mint fails without approval?

Reply:
if you don't approve a token, then the error should be about not having enough allowance.